src/components/forms/forms.js

```diff
@@ -13,80 +21,122 @@ import {
 
 import './forms.css';
 
+export const commonRequiredProps = ['id'];
+export const commonOptionalProps = [
+  'name',
+  'disabled',
+  'description',
+  'title',
+  'ariaLevel',
+  'defaultValue',
+  'hideTitle',
+  'style',
+  'validators',
+  'onChange',
+  'type'
+];
+
+// ----------------------------------------------- //
+// ======       MAIN FORM FIELD TYPES       ====== //
+// ----------------------------------------------- //
 export const FormFieldTypes = {
-  MULTITEXT: { defaultValue: [], component: formInputMultiText },
-  SLIDER: { defaultValue: false, component: formInputSlider },
-  RADIO: { defaultValue: null, component: formInputRadioGroup },
-  TEXT: { defaultValue: '', component: formInputGeneric },
-  NUMBER: { defaultValue: '', component: formInputGeneric },
+  MULTITEXT: {
+    defaultValue: [],
+    component: formInputMultiText,
+    requiredProps: [],
+    optionalProps: [
+      'placeholder',
+      'ariaDescribedby',
+      'inputStyle'
+    ],
+  },
+  SLIDER: {
+    defaultValue: false,
+    component: formInputSlider,
+    requiredProps: [],
+    optionalProps: [
+      'toggleText',
+    ],
+  },
+  RADIO: {
+    defaultValue: null,
+    component: formInputRadioGroup,
+    requiredProps: [
+      'options',
+      // 'options' example:
+      // [
+      //  {name: 'opt_1', text: 'Option 1'},
+      //  {name: 'other', text: 'Other', type: 'string'} <-- will have textbox
+      // ]
+    ],
+    optionalProps: [
+      'ariaDescribedBy',
+    ],
+    customPropValidation: customRadioPropValidation,
+  },
+  TEXT: {
+    defaultValue: '',
+    component: formInputGeneric,
+    requiredProps: [],
+    optionalProps: [
+      'placeholder',
+      'inputStyle',
+      'ariaDescribedby',
+    ],
+  },
+  NUMBER: {
+    defaultValue: '',
+    component: formInputGeneric,
+    requiredProps: [],
+    optionalProps: [
+      'placeholder',
+      'inputStyle',
+      'ariaDescribedby',
+    ],
+  },
   CHECKBOX: {
     defaultValue: (config) => (config?.valueType === 'string' ? '' : false),
```

Comment:
i honestly can't remember, how / when would we want a checkbox to have a string value? If this is a common usecase, would you mind adding a comment in the code for future devs to understand too?

Reply:
I think when the checkbox option is "Other" and to allow the user to input text - came across that when reviewing the Data Governance sub-component in another PR.

Reply:
Yes, Quazi is correct. I will add a comment here + on the field itself; this might be confusing though, and I am open to alternate suggestions.

Reply:
Hmm, I have an idea. What if for both radio and checkbox there was a `conditionalFormField` prop, which, if passed, makes the checkbox enter into a setup where the checkbox controls the conditional rendering of a formField, such as an input box or something else. I suggest this because in another ticket, I need to conditionally render a more complex field than just an input box. Thoughts on this approach?

Reply:
i think any sort of more complex conditional rendering shouldnt be handled as part of the global forms components, and should be handled with the methods that are rendering them.

Instead of `FormField({ conditionalFormField: (condition, template) => ...`, i would rather `FormField({ onChange: () => show/hide stuff`

I think it would help keep the code cleaner and easier to understand long term, since writing all of the code to handle the conditions / checks could get complex, and will largely be unnecessary out of the uses of that one specifically rendered formField

Reply:
I think I agree for checkboxes, though I do think this logic is common enough that having it be a part of the checkbox could be useful. However, for radio groups, this might need to be addressed - since, e.g., there is a radio group in https://github.com/DataBiosphere/duos-ui/pull/1791 which requires conditional rendering - if you check disease specific use, between the disease specific use option and populations, origins, and ancestry option a searchbar needs to come up. In that case, do you think a conditionalFormField would make sense? I am not sure how else to inject that field inside of the radio group.

Reply:
I don't know, it just seems too complex to be nesting formField configs inside formField configs, especially when we can still use `isRendered` and `onChange` handles all of the functionality in the same way. I don't want to be controlling multiple types of objects, or nesting long giant objects, it's harder (for me!) to read.

Interested in seeing what some other opinions are then!

Reply:
I'm with Christine on this. Things like `otherText` (a textbox that shows up if `Other` is selected), can easily be controlled by a different form field with conditional rendering. Those additional fields are often mapped to a variable/key separate from that of the radio button/checkbox.

Take the disease list selection for instance. A similar feature exists in the DAR Form on Step 2. With it, the radio button value is saved to one key, but the follow up value (textbox, list, whatever) is saved to another. The diseases radio button is saved to `diseases` (don't remember the exact name), but the list of diseases itself is saved on `ontologies`

Benefits to this can be applied to the front-end and back-end. For one, the types behind each key are set it stone. A radio button will have to equal a string, a checkbox will have to be true/false, multi-select will have to be an array, etc. If there's a form element down the road that is conditionally rendered on one of these values, I can do a simple equality check vs something like `if(isFunction) {do one thing}, else if(isString) {do something else}...` and so forth. In terms of the back-end, variables of a class need to have their types declared. Having a variable, which is mapped to a checkbox that can assign either a `String` or `Boolean` just adds complexity that the back-end now has to handle.

Reply:
I agree with all of this, but I think functionally speaking this is a bit different than what I was suggesting. Currently, I have the base code pushed up, but I don't have an example of how it would work. Let me flesh out a sketch of how I was intending on using it on the data access governance page:



```
const [diseaseSpecificUseText, setDiseaseSpecificUseText] = useState(...)


h(FormField, {
   type: RADIO,
    ...
   options: [
      ...
      {
         key: "diseaseSpecificUse",
         renderIfSelected: h(FormField, ... form for searching different diseases, onchange updates diseaseSpecificUseText ... )
      }
      ....
   ],
   onChange({}) => {... if (selected == 'diseaseSpecificUse') then setField('diseaseSpecificUse', diseaseSpecificUseText)...},
}

```

This would put a searchbar directly underneath the diesase specific use radio button, and it's on the end developer to do state updates to keep that information. This means the radio operates as one expects, and the user is able to see the searchbar inside of the radio group as they would expect.

Reply:
Approaching this from a high-level view, it's a choice between ...
```
h(FormField), {
  type: RADIO,
  ...
  options: [
    //define other options for parent here
    {
      key: 'diseaseSpecificUse',
      renderIfSelected: h(FormField, {
        type: SELECT,
        ...
        options: [
          //define options for child here
        ],
        onChange: ({})
      })
    }
  ],
  onChange: () => {... if(selected === 'diseaseSpecificUse') then ...}
}
```

and

```
h(FormField), {
  type: SELECT,
  ...
  options: [
    //define options here
  ]
  onChange: () => {/*update parent question value here*/}
},

FormField, {
  type: SELECT,
  ...
  options: [
    //define options here
  ],
  onChange: () => { //update select list here},
  isRendered: diseaseSpecificUse
}
```

Both are truncated definitions, but even so there's one noticeable difference. With the first example the sub-component is now a prop value to the parent component (regardless of whether you attach it to `options` or or pass it in as another prop variable), meaning that if any value on the parent prop is updated, the entire component is re-rendered. Since the child component is a part of the parent, its forced to re-render as well. If the child has any specific logic that fires on initialization, it will fire off once again. 

You can work around the above situation by using a `useMemo` hook to define the child component so that it re-renders on specific variable updates, but now you're in a situation where you've defined two components (with one passed in as a child component to the other) alongside a useMemo variable.

The second option doesn't have to worry about re-rendering outside of the `isRendered` condition and input updates. If the `options`, `onChange`, `disabled`, or any prop on the non-dependent (parent) component changes, the dependent (child) component will be unaffected.

So then the question now is whether or not there's any significant gains to writing the code this way. Reasons could be anything from performance, testability, organization, readability, etc. From a high-level perspective I don't see any noticeable differences between version 1 and version 2 other than having the child component nestled under the parent in a container div. In terms of readability and performance version 1 takes a hit because we now have to deal with the complexity of nested components, which is why I lean towards version 2. If there was any additional complexity then I could lean towards 1, but right now we're just accomplishing the same effect in version 2 with more overhead instead of less.

Reply:
I totally agree with what you're saying about complexity, and if they rendered the exact same HTML then I would prefer the second option. However, the reason I am pushing for this option specifically is becuase the mock looks like so:

<img width="563" alt="image" src="https://user-images.githubusercontent.com/17440010/189982057-ac2aef6c-986a-43e4-ae7b-3f7769879bcc.png">

I think for the end user having the searchbar be inside of the radio group is a better look for the option. If we don't nest the searchbar inside of the radio group's formfield, then having the search bar be a seamless part of the radio group isn't possible. If we're okay with putting the searchbar below the radio group, then I totally agree and will remove this functionality; however, most of this work has been in effort to figure out how to accomplish the nested searchbar.

Reply:
Ah the form component is a template for the entire group, not the individual component. The part I'm having trouble rectifying is this line

``` If we don't nest the searchbar inside of the radio group's formfield, then having the search bar be a seamless part of the radio group isn't possible```

The radio buttons don't have to exist as part of a group template, they can exist as individual elements. UI libraries and company style guides will always develop individual `RadioButton` components explicitly for edge cases where you need custom formats and layouts. This way you can style and arrange them as you see fit for the page in question. You want the search bar/input/textfield to appear above, left, right, below? Easy to handle if the elements are drawn out individually in HTML. But if it's passed in as a prop? To handle the custom positioning you'll need even more custom props.

This isn't to say that the Group template is bad, it's still useful for situations where there are no extra fields popping in and out. But for custom layouts the accepted approach is to manually write out the fields.

Here's Atlassian's documentation on radio buttons:
https://atlassian.design/components/radio/examples

Here's Material-UI's take on radio buttons (standalone)
https://mui.com/material-ui/react-radio-button/#standalone-radio-buttons

Here's Material-UI's approach on RadioGroups
https://mui.com/material-ui/react-radio-button/#radio-group

Reply:
Ok this these are all great points. I think I will remove this functionality so that this PR can get merged in. In my `DataAccessGovernance` PR I will work on a FormField which is just a singular radio button.